api/src/main/java/org/openmrs/module/referencedemodata/ReferenceDemoDataActivator.java

```diff
@@ -206,6 +206,7 @@ public void started() {
 					
 					try {
 						Context.flushSession();
+						Context.clearSession();
```

Comment:
Do you mind throwing some more light on how, clearing cached changes without writing them to the database, improves performance?

Reply:
`Context.flushSession();` should persist any unsaved cached changes. `Context.clearSession()`  should release memory that the app should otherwise utilise hence boosting its performance.

Reply:
Makes sense now. Thanks :)

Reply:
I'm a little wary of adding this `clearSession()`, but because this will break the cache mechanism. I suppose if we also have a call to flush to internal object cache this is fine.

Reply:
Isn't this already done on line 215?

Reply:
Doing this fairly early on every patient whose demo-data has been generated would be ideal unlike doing it at once on line 215 when all demo-data has been generated.

Reply:
Right, but to do that you'd have to clear the session inside the loop... and if you do that, you'll also have to clear the cache since otherwise the cache will end up holding onto Hibernate objects that are no longer associated with the session properly.

But if the concern here is performance, you've now changed things to make two function calls where one will suffice.

Reply:
Ahh, ok. But I think `clearSession()` is indeed clearing the hibernate cache. Or by cache could you be referring to the `conceptCache`, IMO clearing them from the hibernate catch shouldn't erase their `id`s.

Reply:
I'm referring to the concept cache... You're right that `clearSession()` clears the Hibernate cache. It doesn't erase their ids, but it does move them into an invalid state.

In order to keep things synchronised, all Hibernate objects[^1] have a reference to the `session` they were loaded in. Clearing the Hibernate cache removes the copy of them from the session, but has no effect on the object itself per-se. The problem comes when we try to use an object, e.g.,

```
Obs.setConcept(someCachedConcept);
```

In the above line, after the session has been cleared, `someCachedConcept` now refers to an invalid object (because it's associated with a session that is no longer storing the state of that object) and so the transaction will be rejected. That's why I moved the call to `clearSession()` outside of the loop.

In either case, the point I was trying to make is that having the call here has the same effect as having the call on line 215 and we don't need both of these calls... The one on line 215 should be preferred though, as it happens regardless of whether an exception gets thrown or not.

[1]: Hibernate does have the notion of a "detached" object, but detaching and reattaching an object are operations that need to be called by the code. They don't happen automatically.

Reply:
Ok @ibacher. Thanks alot for the detailed explanation. I'll undo this line then, to leave everything consistent/synchronised with hibernate. It does add ~3 minutes averagely to the loading time which is tolerable.